File: frontend/packages/types/src/learning.ts
// ── Learning 域类型（课程生成 + 进度追踪）────────────────────────────────────
//
// 与 `backend/app/schemas/learning.py` 对齐（task-337 后端契约）。
// 网关 `@readinglist/api/gateways/learning` 负责解包 `APIResponse` 信封，
// 消费方拿到的就是这里的形状。

/** 选择题单选题选项 */
export interface ExerciseOption {
  key: string;
  text: string;
}

/** 题型字面量 */
export type ExerciseType = 'single_choice' | 'multi_choice' | 'true_false';

/** 课程生成状态字面量 */
export type CourseStatus = 'pending' | 'ready' | 'failed';

/** 单个练习题（来自 exercise.md 解析） */
export interface Exercise {
  id: number;
  type: ExerciseType;
  difficulty: number;
  points: number;
  prompt: string;
  /** 选择题必有；判断题恒为 `null` */
  options: ExerciseOption[] | null;
  /** 单选 `string` / 多选 `string[]` / 判断 `boolean` */
  answer: string | string[] | boolean;
  explanation: string;
}

/** 单个课程（teach skill 对齐：一课一文件） */
export interface LearningLesson {
  id: number;
  title: string;
  slug: string;
  /** 该课正文全文 */
  md: string;
  /** 该课练习（解析自 `<num>-<slug>.exercise.md` 的 front matter） */
  exercises: Exercise[];
}

/** 完整课程包（`ready` 时随 `GET /courses/{id}` 一起下发） */
export interface LearningCourse {
  course_id: string;
  topic: string;
  /** 已生成的课（渐进产出时列表增长） */
  lessons: LearningLesson[];
  resource_md: string;
  /** 学习使命文档（MISSION.md 全文）；缺失（旧课程 / 未生成）为 `null` */
  mission_md: string | null;
}

/** 课程进度项（来自 `GET /v2/learning/progress`） */
export interface LearningProgressItem {
  course_id: string;
  topic: string;
  sessions_done: number[];
  exercise_done: boolean;
  status: CourseStatus;
  next_session: number | null;
}

/** `GET /v2/learning/courses/{course_id}` 的按状态判别的响应 */
export type CourseStatusResponse =
  | { status: 'pending'; course_id: string }
  | { status: 'ready'; course: LearningCourse }
  | { status: 'failed'; course_id: string };

/** `POST /v2/learning/courses/{course_id}/lessons` (task-352) 的响应字面量。 */
export type NextLessonStatus = 'pending' | 'already_generated' | 'failed';

/**
 * 单个可用学习模型（来自 `GET /v2/learning/models`，task-391）。
 *
 * `is_premium === true` 时前端应在未登录状态下禁用选项；登录用户可正常
 * 选择。后端白名单 (`LEARNING_MODELS`) 与 generator 同源，前端无需单独
 * 维护常量列表。
 */
export interface LearningModel {
  /** 后端白名单 key，例如 `deepseek-v4-flash` / `deepseek-v4-pro`。 */
  id: string;
  /** 展示用文案（任务 391 后端已对齐 i18n key，前端直接显示）。 */
  label: string;
  /** 仅登录用户可用的模型 — 未登录用户禁选并提示登录。 */
  is_premium: boolean;
}

/** `POST /v2/learning/courses/{course_id}/lessons` 的响应。 */
export interface NextLessonResponse {
  course_id: string;
  /** 预期/已生成的新课编号；幂等命中或失败时为 `null`。 */
  next_lesson: number | null;
  status: NextLessonStatus;
}

File: backend/app/schemas/learning.py
"""Learning schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

# ── 课程生成与产物 ───────────────────────────────────────────────────────── #


class CourseGenerateInput(BaseModel):
    """用户提交的学习主题"""

    topic: str = Field(..., min_length=1, description="要学习的主题")
    goal: str | None = Field(
        default=None,
        max_length=200,
        description="学习目标（可选，≤200 字符），用于组织 MISSION.md 具体文案",
    )


class CourseStatus:
    """课程生成状态字面量集合。"""

    PENDING: Literal["pending"] = "pending"
    READY: Literal["ready"] = "ready"
    FAILED: Literal["failed"] = "failed"


CourseStatusLiteral = Literal["pending", "ready", "failed"]


# ── 练习（Exercise） ───────────────────────────────────────────────────── #


class ExerciseOption(BaseModel):
    """选择题选项"""

    key: str = Field(..., description="选项键，如 A/B/C/D")
    text: str = Field(..., description="选项文案")


class Exercise(BaseModel):
    """单个练习题"""

    id: int = Field(..., ge=1, description="题号，从 1 开始")
    type: Literal["single_choice", "multi_choice", "true_false"] = Field(
        ..., description="题型"
    )
    difficulty: int = Field(..., ge=1, le=3, description="难度（1–3）")
    points: int = Field(..., ge=0, description="分值")
    prompt: str = Field(..., description="题干")
    options: list[dict[str, str]] | None = Field(
        default=None,
        description="选项列表 [{key, text}]，选择题必填，判断题留空为 None",
    )
    answer: str | list[str] | bool = Field(
        ..., description="参考答案：单选 str / 多选 list[str] / 判断 bool"
    )
    explanation: str = Field(..., description="解答说明（必填，D2）")


# ── 课程包 ──────────────────────────────────────────────────────────────── #


class LessonItem(BaseModel):
    """单个课程（teach skill 对齐：一课一文件）。

    ``id`` 从 1 起，是进度 ``sessions_done`` 的编号依据；``slug`` 参与
    磁盘文件名（``<num>-<slug>.md`` 与 ``<num>-<slug>.exercise.md``）。
    """

    id: int = Field(..., ge=1, description="课序号，从 1 开始")
    title: str = Field(..., description="课标题")
    slug: str = Field(..., description="dash-case 课标识，用于文件名")
    md: str = Field(..., description="该课正文全文")
    exercises: list[Exercise] = Field(
        default_factory=list,
        description="该课练习（解析自 <num>-<slug>.exercise.md 的 front matter）",
    )


class CoursePackage(BaseModel):
    """按课重构后的课程包（内存层结构）。

    - ``lessons`` 只含**已生成**的课（渐进产出时列表随 ``POST /lessons`` 增长）。
    - ``resource_md`` 是全课程共享参考资料，独立于各课。
    - ``mission_md`` 是本课程学习使命文档（MISSION.md 全文），缺失为 None。
    """

    course_id: str = Field(..., description="课程唯一 ID")
    topic: str = Field(..., description="原始主题")
    lessons: list[LessonItem] = Field(
        default_factory=list, description="已生成的课列表（渐进产出时增长）"
    )
    resource_md: str = Field(..., description="resource.md 全文")
    mission_md: str | None = Field(
        default=None, description="MISSION.md 全文（学习使命文档），缺失为 None"
    )
